src/main/java/org/candlepin/resource/EnvironmentResource.java

```diff
@@ -255,8 +260,21 @@ private void deleteConsumers(Environment environment, List<Consumer> consumers)
     }
 
     @Override
-    public Iterable<EnvironmentDTO> getEnvironments() {
-        return translator.translateQuery(this.envCurator.listAll(), EnvironmentDTO.class);
+    public Stream<EnvironmentDTO> getEnvironments(List<String> type, Boolean listAll) {
+        List<Environment> envs;
+        if (listAll != null && listAll) {
+            envs = this.envCurator.listAllIgnoreType();
+        }
+        else if (type == null || type.size() == 0) {
+            envs = this.envCurator.listAllNullType();
+        }
+        else {
+            envs = this.envCurator.listAllByType(type);
+        }
+        Stream<EnvironmentDTO> stream = envs.stream()
+            .map(this.translator.getStreamMapper(Environment.class, EnvironmentDTO.class));
+        return this.pagingUtilFactory.forClass(EnvironmentDTO.class)
+            .applyPaging(stream, envs.size());
```

Comment:
Up until now, only satellites used environments, and with single-digit amount of environments per org, and single-digit amount of orgs in a satellite, this API would never return more than double-digit amount of environments (realistically, no more than 25 in extreme cases), so this would be fine.

However, with this feature, environments are now in Hosted, which has around ~133k active owners (millions total), and it is rising every month. Assuming even only a fifth of these active owners use this feature eventually, and an average of 2 environments per owner, this API will fetch ~50k environment instances into memory every time its called and paginate after the fact, which is not good.

Instead of the PagingUtilFactory we need to do paging at the DB query level, similarly to `JobResource.listJobStatuses` and `ConsumerResource.searchConsumers` (and limit the amount of results returned if no paging was requested by throwing an error after a reasonable amount of results would be returned).

The owners/owner_key/environments can use the pageutil because it won't even return enough environments to be an issue

Reply:
Also, looking at Katello code, they don't use this endpoint, and looking at the HMS content-sources list of endpoints they want access to, they're not going to use this endpoint either.

Which makes it a good candidate for removal, or alternatively leaving it for debugging purposes.
@Ceiu @wottop what do you think?

Reply:
Given how environments are used, the "get me all environments globally" query is one of those ones that didn't make sense to me, but I didn't look into it at all. If no one is using it, I'm down for just removing it.

If we opt to keep it, Nikos' assessment of the potential data load is fair and I would agree that we should apply more thorough paging.

Reply:
Personally I don't see how this endpoint would be helpful for debugging, so lets remove it.